exla/lib/exla/mlir/value.ex

```diff
@@ -434,206 +542,414 @@ defmodule EXLA.MLIR.Value do
+    {token, results}
+  end
 
-    {results, %Region{ref: true_region}, %Region{ref: false_region}}
+  def outfeed(%Value{} = input, token), do: outfeed([input], token)
+
+  def outfeed(inputs, %Value{function: func} = token) do
+    result_types = [type_token()]
+    op(func, "stablehlo.outfeed", inputs ++ [token], result_types) |> hd()
+  end
+
+  def create_token(%Function{} = func) do
+    result_types = [type_token()]
+    op(func, "stablehlo.create_token", [], result_types) |> hd()
   end
 
-  def infeed(%Value{} = token, %EXLA.Shape{} = shape) do
-    infeed(token, [shape])
+  def call(%Function{} = func, args, %Function{} = computation, typespecs) do
+    result_types = typespecs_to_mlir_types(typespecs)
+    attributes = [callee: attr_symbol_reference(computation.name)]
+    op(func, "func.call", args, result_types, attributes: attributes)
   end
 
-  def infeed(%Value{function: function} = token, [%EXLA.Shape{} | _] = shapes) do
-    {token_ref, result_refs} =
-      EXLA.NIF.mlir_infeed(function.ref, token.ref, Enum.map(shapes, & &1.ref)) |> unwrap!()
+  def while(%Function{} = func, %Region{ref: pred}, %Region{ref: body}, initial) do
+    typespecs = Enum.map(initial, &get_typespec/1)
+    result_types = typespecs_to_mlir_types(typespecs)
 
-    {%Value{token | ref: token_ref}, Enum.map(result_refs, &%Value{token | ref: &1})}
+    regions = [pred, body]
+
+    op(func, "stablehlo.while", initial, result_types, regions: regions)
   end
 
-  def outfeed(%Value{} = input, token), do: outfeed([input], token)
+  def return(func, values) when is_list(values) do
+    op(func, "stablehlo.return", values, [])
+  end
+
+  def qr(%Value{function: func} = value, q_typespec, r_typespec) do
+    %{type: op_type, shape: op_shape} = get_typespec(value)
+    %{type: q_type, shape: q_shape} = q_typespec
+    %{type: r_type, shape: r_shape} = r_typespec
+
+    dim_sizes = [tuple_size(op_shape), tuple_size(q_shape), tuple_size(r_shape)]
+    operand_dims = Tuple.to_list(op_shape)
+    q_dims = Tuple.to_list(q_shape)
+    r_dims = Tuple.to_list(r_shape)
+
+    dim_sizes = constant(func, dim_sizes, Typespec.tensor({:s, 64}, {length(dim_sizes)}))
+    operand_dims = constant(func, operand_dims, Typespec.tensor({:s, 64}, {length(operand_dims)}))
+    q_dims = constant(func, q_dims, Typespec.tensor({:s, 64}, {length(q_dims)}))
+    r_dims = constant(func, r_dims, Typespec.tensor({:s, 64}, {length(r_dims)}))
+    operands = [value, dim_sizes, operand_dims, q_dims, r_dims]
+
+    q_result_type = type_tensor(q_type, q_shape)
+    r_result_type = type_tensor(r_type, r_shape)
+    result_types = [type_tuple([q_result_type, r_result_type])]
+
+    call_target_name =
+      case op_type do
+        {:f, 32} -> "qr_cpu_custom_call_f32"
+        {:f, 64} -> "qr_cpu_custom_call_f64"
+        {:f, 16} -> "qr_cpu_custom_call_f16"
+        {:bf, 16} -> "qr_cpu_custom_call_bf16"
+        type -> raise "QR decomposition not supported for type #{inspect(type)}"
+      end
```

Comment:
This function should be receiving the `device` for asserting that it is the host. Otherwise the error message is more confusing than it should be.

Reply:
We already match on `:host` here:

https://github.com/elixir-nx/nx/blob/3b3e92ebc733fb0961813679618f608ee501c199/exla/lib/exla/defn.ex#L580-L585

Reply:
Yes, I know, but the user doesn't know that. And then the error message hides this implementation detail

Reply:
Actually, this is an optional, so if the clause doesn't match it falls back to the optional implementation. I think it's all good, let me know otherwise!